# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/loader/ModuleLoaderBase.cpp
# Commit: a1900e8b640c
# Full Hash: a1900e8b640c845614da76612f92b76b3f199a75
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2023-01-19 04:46:07
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Handle cancellation of long running modules; r=jonco
#   
#   This is a slightly annoying thing that can happen. When we abruptly cancel (such as an infinitely
#   running script being forcibly terminated) we will be in a state where the EvaluateModule call will
#   finish _after_ the loader is destroyed. So, instead we track if there has been a forcible
# ==============================================================================

diff -r c585aea29f0c -r a1900e8b640c js/loader/ModuleLoaderBase.cpp
--- a/js/loader/ModuleLoaderBase.cpp	Wed Jan 18 13:46:32 2023 +0000
+++ b/js/loader/ModuleLoaderBase.cpp	Wed Jan 18 13:46:32 2023 +0000
@@ -1212,6 +1212,12 @@
   // unless the user cancels execution.
   MOZ_ASSERT_IF(ok, !JS_IsExceptionPending(aCx));
 
+  // For long running scripts, the request may be cancelled abruptly. This
+  // may also happen if the loader is collected before we get here.
+  if (request->IsCanceled() || !mLoader) {
+    return NS_ERROR_ABORT;
+  }
+
   if (!ok) {
     LOG(("ScriptLoadRequest (%p):   evaluation failed", aRequest));
     // For a dynamic import, the promise is rejected. Otherwise an error is
